### ISSUE TITLE
Consider implementing a syntax-aware structured diff implementation

### ISSUE THREAD

**Issue by krishnakumarg1984** [neovim/neovim#15064]
I love `neovim` and have been using it since 2017 (just prior to version 0.2).  The features introduced in the latest 0.5 release (lua, LSP, treesitter etc.) make it even more awesome.

I'd like to request the devs to consider a structure-based diff implementation (to complete the existing suite of line-based diffing algorithms already present i.e. Myers, patience, longest subsequence distance, histogram etc.). I recently came across [`difftastic`](https://github.com/Wilfred/difftastic), a structure-based diffing tool (even though this is currently a rough, unfinished product). To my untrained eye, structure-based diffing looks like a nice orthogonal idea that is worthwhile to implement. The creator of difftastic further points out some interesting structure-based diff tools in the project's [wiki page](https://github.com/Wilfred/difftastic/wiki/Structural-Diffs).

Here are some further related readings/tools (originally linked in the aforementioned wiki):

1. [https://blog.trailofbits.com/2020/08/28/graphtage/](https://blog.trailofbits.com/2020/08/28/graphtage/)
2. [https://fosdem.org/2021/schedule/event/sexpressiondiff/](https://fosdem.org/2021/schedule/event/sexpressiondiff/)
3. [https://fazzone.github.io/autochrome.html](https://fazzone.github.io/autochrome.html)
4. [https://github.com/yinwang0/psydiff](https://github.com/yinwang0/psydiff)
5. [https://github.com/yinwang0/ydiff](https://github.com/yinwang0/ydiff)
6. [https://github.com/VictorCMiraldo/hdiff/tree/v0.0.5](https://github.com/VictorCMiraldo/hdiff/tree/v0.0.5)
7. [https://github.com/GumTreeDiff/gumtree](https://github.com/GumTreeDiff/gumtree)

I gather that structure-based diffing might be a reasonably hot topic in academia, with a few good PhD theses and papers being published e.g. [https://victorcmiraldo.github.io/data/MiraldoPhD.pdf](https://victorcmiraldo.github.io/data/MiraldoPhD.pdf). 

I am just a user of neovim, often using it for writing notes, papers and application-level code, and do not have the background in data structures required to implement a sophisticated idea like this. I hope that the community here might be able to have a healthy debate on the pros and cons of considering a structure-based diff implementation within neovim.

**Comment by bfredl** [comment]
https://github.com/afnanenayet/diffsitter this also looks interesting as it is based on tree-sitter.

**Comment by clason** [comment]
Yes, diffsitter looks like the way to go here. It shouldn't be too hard to port that to an [nvim-treesitter module](https://github.com/nvim-treesitter/nvim-treesitter#adding-modules) (starting from https://github.com/nvim-treesitter/module-template), which would be the right place for this at the moment.

**Comment by krishnakumarg1984** [comment]
One thing that is not clear from diffsitter's README.md page is whether it supports a nice-looking split view? All1 the examples in that page return classical diff markers.

**Comment by clason** [comment]
We wouldn't use diffsitter directly but use the same algorithm to generate a diff using tree-sitter. The display would still use the standard vim functions (although they may have to be adapted to account for such different files being marked identical).

**Comment by matu3ba** [comment]
It is not sufficient to use the more optimal [graph-based implementation](https://github.com/Wilfred/difftastic), because "Performance. Difftastic scales poorly on files with a large number of changes. This might be solved by A* search."

[A* is quite good explained here](https://thume.ca/2017/06/17/tree-diffing/).
So either difftastic includes A* (or another performant algorithm on too many individual changes) or user experience will suffer on code that others write (and tree-sitter can not recommend when to commit changes).

**Comment by lewis6991** [comment]
No interest from me, and tbh this can easily be implemented via a plugin. There's no point adding this directly in core when the premise/utility of the feature is unproven.

Let's demonstrate the concept via a plugin first, then we can assess whether it's worth baking this in.

**Comment by theHamsta** [comment]
Mentioning https://github.com/Wilfred/difftastic for another approach to implement this as CLI tool.

**Comment by krishnakumarg1984** [comment]
Well my OP mentions `difftastic`.

**Comment by hacker-DOM** [comment]
Fwiw, `difftastic` is also tree-sitter-based. It looks like the most developed and advanced of these structural differs

**Comment by shelper** [comment]
so any way to integrate difftastic into neovim's diff view? e.g., gitsigns diffthis

**Comment by franroa** [comment]
I was looking for the same
